### PR TITLE
Ensure that a StreamNode is always owned by only one StreamBitmap

### PR DESCRIPTION
**Pull request by armenatzoglou**
in a `Bitmap Table Scan`, there are cases where more than one bitmaps point to the same `StreamNode`. Thus, when we free the bitmaps and the `StreamNodes` that they point, we end up to a double free of a `StreamNode`.

This is a repro:
```
set optimizer=on;
create table foo (c int, d int) distributed by (c);
create index ie_foo on foo using bitmap(d);
insert into foo values (1, 1);
insert into foo values (2, 2);

explain select * from foo where d = 2 and (d = 1 or d = 2);
                                      QUERY PLAN
---------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..0.09 rows=1 width=8)
   ->  Bitmap Table Scan on foo  (cost=0.00..0.09 rows=1 width=8)
         Recheck Cond: d = 2 AND (d = 1 OR d = 2)
         ->  BitmapAnd  (cost=0.00..0.00 rows=0 width=0)
               ->  Bitmap Index Scan on ie_foo  (cost=0.00..0.00 rows=0 width=0)
                     Index Cond: d = 2
               ->  BitmapOr  (cost=0.00..0.00 rows=0 width=0)
                     ->  Bitmap Index Scan on ie_foo  (cost=0.00..0.00 rows=0 width=0)
                           Index Cond: d = 1
                     ->  Bitmap Index Scan on ie_foo  (cost=0.00..0.00 rows=0 width=0)
                           Index Cond: d = 2
 Settings:  optimizer=on
 Optimizer status: PQO version 2.1
(13 rows)

select * from foo where d = 2 and (d = 1 or d = 2);
```

In the above scenario, when GPDB frees `BitmapAND`, it will also free bitmap's associated `StreamNode` and all latter's input streams that include the `StreamNode` of OR. Consequently, when we free `BitmapOR`, GPDB will also attempt to free OR's `StreamNode`. Thus, we have a double free. 

In this PR, we change the ownership of a `StreamNode`, so that when a `StreamNode` A is added as an input to a `StreamNode` B, the bitmap that was pointing to A abandons the ownership and the pointer is set to NULL. In the above example, when GPDB frees BitmapOR, it will not attempt to free OR's `StreamNode`. 

In addition, if a `StreamNode` is an `OpStream`, then `opstream_free` function is used to free the memory used by the `StreamNode`, which actually uses `pfree`. On the other hand, if a StreamNode is an IndexStream, then `stream_free` function is invoked, which does not pfree the StreamNode. For consistency, we add function `indexstream_free` to always pfree a `StreamNode` of `IndexStream` type.

### CODE REVIEW

**Comment by armenatzoglou** [comment]
@foyzur @karthijrk @xinzweb Please take a look when you have some free time.

**Comment by foyzur** [comment]
Where are we freeing the AND/OR bitmaps? I see that AND/OR claimed ownership of the underlying stream node, but I don't see anywhere where we free such bitmap. Now, that being said, if the container bitmap comes from a BitmapIndexScan, then the ExecEndBitmapIndexScan would free that eventually. But that has two catches:

1. This free is ignored for planner plan
2. For some cases AND/OR will construct their own StreamBitmap and the BitmapIndexScan will not have access to such StreamBitmap and so it won't be able to free that bitmap.

Am I missing something?

**Comment by foyzur** [comment]
Also, I don't see any test that actually verifies that we are not leaking memory. I would suggest a large number of rescan of the bitmap (i.e., index join on the inner side of NLJ, e.g.) and then compare the peak memory to see if it looks reasonable. Also, such peak shouldn't change much depending on the outer row count.

File: src/backend/executor/nodeBitmapAnd.c [hunk 1]
```diff
@@ -203,6 +211,8 @@ MultiExecBitmapAnd(BitmapAndState *node)
 	   				StreamBitmap *s = (StreamBitmap *)subresult;
 		   			stream_add_node((StreamBitmap *)node->bitmap,
 									s->streamNode, BMS_AND);
+		   			/* node->bitmap should be the only owner of the newly created AND StreamNode */
+		   			s->streamNode = NULL;
```
Comment (karthijrk) [inline src/backend/executor/nodeBitmapAnd.c]: Is there a scenario where we need to share the same stream node to multiple Stream Bitmap?

File: src/backend/executor/nodeBitmapAnd.c [hunk 2]
```diff
@@ -203,6 +211,8 @@ MultiExecBitmapAnd(BitmapAndState *node)
 	   				StreamBitmap *s = (StreamBitmap *)subresult;
```
Comment (karthijrk) [inline src/backend/executor/nodeBitmapAnd.c]: Who owns this subresult ? Who will call free on subresult that would result in free of subresult->streamNode ?

**Comment by karthijrk** [comment]
Thanks Nikos for explaining in detail.

**Comment by karthijrk** [comment]
I have the same concern as Foyzur. Otherwise it LGTM :)

**Comment by armenatzoglou** [comment]
> Where are we freeing the AND/OR bitmaps? I see that AND/OR claimed ownership of the underlying stream node, but I don't see anywhere where we free such bitmap. Now, that being said, if the container bitmap comes from a BitmapIndexScan, then the ExecEndBitmapIndexScan would free that eventually. But that has two catches:

> 1. This free is ignored for planner plan

Indeed, in ExecEndBitmapIndexScan, if the plan was generated by the planner then we do not free the bitmap (see below). 

```
/* Only for optimizer BitmapIndexScan is in charge to free the bitmap */
	if(PLANGEN_OPTIMIZER == node->indexScanState.ss.ps.state->es_plannedstmt->planGen)
	{
		tbm_bitmap_free(node->bitmap);
	}
	node->bitmap = NULL;
```

However, this code was introduced by you (see commit 3e286be82ce306e1ae5b92270a455accdae1192b) and it is not related to this issue/PR. 

> 2. For some cases AND/OR will construct their own StreamBitmap and the BitmapIndexScan will not have access to such StreamBitmap and so it won't be able to free that bitmap.

As per our discussion, there are no AND/OR bitmaps. In particular, when we AND (or OR) two bitmaps A and B, we i) merge B inside A, ii) create a new StreamNode of type BMS_AND (or BMS_OR), iii) set A to point to the newly created StreamNode, and iv) add A->streamnode and B->streamnode as children of BMS_AND (or BMS_OR) StreamNode. Thus, there is no concern.

File: src/backend/executor/nodeBitmapAnd.c (hunk 1 above)
Comment (armenatzoglou) [inline src/backend/executor/nodeBitmapAnd.c]: To the best of my knowledge, there is no such a scenario.

File: src/backend/executor/nodeBitmapAnd.c (hunk 2 above)
Comment (armenatzoglou) [inline src/backend/executor/nodeBitmapAnd.c]: As you mentioned in the conversation page, we already discussed that.

**Comment by armenatzoglou** [comment]
@foyzur 
> Also, I don't see any test that actually verifies that we are not leaking memory. I would suggest a large number of rescan of the bitmap (i.e., index join on the inner side of NLJ, e.g.) and then compare the peak memory to see if it looks reasonable. Also, such peak shouldn't change much depending on the outer row count.

We agree that a test for memory consumption is usefull here. This is not a memory leak because all memory is allocated in ExecutorState memory context, so we know that it will be freed correctly. 

Also, in https://github.com/greenplum-db/gpdb/commit/3e286be82ce306e1ae5b92270a455accdae1192b you added a banch of tests that examine if rescan of a bitmap works as expected. So, we can reuse them. However, as per our discussion, having a test that checks if the peak of the memory usage of a BitmapAnd/BitmapOr exceeds a specific constant (i.e., comment out the free process and get the memory usage) is not deterministic, since it depends on several environment setings, e.g., number of segments. You can find the test here: https://github.com/armenatzoglou/gpdb/commit/a8381f44fb2fbf2bffc1698aeaf763edd45e95dd

We prefer not to add such a test. Instead, we have created a story in QX's backlog to investigate an efficient and viable approach to test GPDB for memory leaks (story id: #138205749). In this fix we are safe because i) all StreamNodes are allocated in ExecutorState memory context, ii) all tests that we added in this PR are green, and iii) the test in https://github.com/armenatzoglou/gpdb/commit/a8381f44fb2fbf2bffc1698aeaf763edd45e95dd passes locally too.

**Comment by armenatzoglou** [comment]
Pushed to master 9dfd0535ffb124f68145913fe8a7bce5f3b4e231. Thanks a lot for the review.

**Pull request closed without merging**